system/core/document.php

```diff
@@ -568,11 +569,52 @@ public function dataFor($name, $default = null) {
 			return $default;
 		}
 
-		function validateRequiredField($name) {
-			if (!array_key_exists($name, $this->fields)) {
-				$this->errors[$name] = __('field-not-found');
-				return false;
+		/**
+		 * Is this form valid? Returns true when:
+		 *  - validate() was called
+		 *  - all of the fields for which validateXXXField
+		 *    methods were called were also valid.
+		 */
+		function isValid() {
```

Comment:
Executing the form validate() here, if not yet done, would reduce some code duplication. It would loose it's const-ness but it would perform more as the name suggests.

Reply:
Interesting idea, though I'm not entirely sure. Always running validate would mean a form can be validated more than once (which is a waste of efffort), but running it only if not ran before means that re-validating after a change in the form is not possible (we don't really use or support this now, though).

But maybe running it when not ran before is a good idea. If we ever need to re-validate, you could just call `validate()` explicitly then. I'll see how to best implement this, since it might nicest to fold this change back into the git history (but maybe that's too much work and I'll just add a new commit).

Reply:
If it would self validate, then, when a value gets updated, the is_validated flag could be set to false.

Reply:
Ok, I implemented automatic validation like suggested. Re-validating / clearing `is_validated` on changes is not yet implemented, since there is no meaningful way to change values right now I believe, so this is left for later.